### PR TITLE
Filter Actions workflows by path

### LINKED ISSUE #3735: GitHub Actions jobs - filter by path?
Actions allows filtering jobs by path.  See:
https://help.github.com/en/actions/reference/workflow-syntax-for-github-actions#onpushpull_requestpaths

There are an awful lot of jobs being run, and the workflows don't seem to indicate that all RG jobs need to run for a Bundler change, and vice versa.

So, I see three types of jobs:
1. Jobs that only need to run with a RubyGems change
2. Jobs that only need to run with a Bundler change
3. Jobs that need to run with any change

With some reorganization, this could be corrected.  It may require splitting jobs in workflows, or adding additional workflows, but it would quicken CI time quite a bit...

Thoughts?

I will abide by the [code of conduct](https://github.com/rubygems/rubygems/blob/master/CODE_OF_CONDUCT.md).

### PR DESCRIPTION
**Pull request by MSP-Greg**
# Description:

Very few of the Actions CI jobs need to run if a commit/PR affects either Bundler or RubyGems files.  Most jobs need to be run only if one or the other changes.

## What was the end-user or developer problem that led to this PR?

CI takes too long, with unneeded jobs being run.

## What is your fix for the problem, implemented in this PR?

Using Actions `paths:` and `paths-ignore:`, workflows can be ran based on which files have changed in a PR/commit.  Use filters `"bundler/**"` and `".github/workflows/*bundler*.yml"` to positive filter for Bundler jobs, and use them with negative filter for RubyGems jobs.

Added a push branch filter of `"pr-**"`, so contributors can run CI in the forks, as long as the branch starts with `pr-`. Presently, the only way CI runs in a fork is from the `master` branch.

Closes #3735
______________

# Tasks:

- [x] Describe the problem / feature
- [x] Ran tests
- [x] Write code to solve the problem
- [ ] Get code review from coworkers / friends

I will abide by the [code of conduct](https://github.com/rubygems/rubygems/blob/master/CODE_OF_CONDUCT.md).

### CODE REVIEW

**Comment by MSP-Greg** [comment]
I forgot to filter with `.rubocop_bundler.yml`.  I'll add it...

File: .github/workflows/install.yml [hunk 1]
```diff
@@ -6,6 +6,7 @@ on:
   push:
     branches:
       - master
+      - "pr-**"
```
Comment (deivid-rodriguez) [inline .github/workflows/install.yml:9]: Please remove all these changes. I really appreciate your contributions, but I'm really tired of telling you to remove unrelated stuff from your PRs :(

File: .github/workflows/jruby-bundler.yml [hunk 2]
```diff
@@ -2,10 +2,17 @@ name: jruby-bundler
...
 on:
   pull_request:
+    paths:
+      - "bundler/**"
+      - ".github/workflows/*bundler*.yml"
 
   push:
     branches:
       - master
+      - "pr-**"
```
Comment (deivid-rodriguez) [inline .github/workflows/jruby-bundler.yml:12]: :scissors:

File: .github/workflows/windows-bundler.yml [hunk 3]
```diff
@@ -16,7 +23,7 @@ jobs:
 
     strategy:
       matrix:
-        ruby: [ '2.4', '2.5', '2.6', '2.7' ]
+        ruby: [ 2.4, 2.5, 2.6, 2.7 ]
```
Comment (deivid-rodriguez) [inline .github/workflows/windows-bundler.yml:26]: :scissors:

File: .github/workflows/windows.yml [hunk 4]
```diff
@@ -22,7 +29,6 @@ jobs:
           bundler: none
       - name: Install Dependencies
         run: rake setup
-        shell: bash
```
Comment (deivid-rodriguez) [inline .github/workflows/windows.yml:25]: :scissors:

File: .github/workflows/windows.yml [hunk 5]
```diff
@@ -2,10 +2,17 @@ name: windows
...
 on:
   pull_request:
+    paths-ignore:
+      - "bundler/**"
+      - ".github/workflows/*bundler*.yml"
 
   push:
     branches:
       - master
+      - "pr-**"
```
Comment (deivid-rodriguez) [inline .github/workflows/windows.yml:12]: :scissors:

File: .github/workflows/windows-bundler.yml [hunk 6]
```diff
@@ -2,10 +2,17 @@ name: windows-bundler
...
 on:
   pull_request:
+    paths:
+      - "bundler/**"
+      - ".github/workflows/*bundler*.yml"
 
   push:
     branches:
       - master
+      - "pr-**"
```
Comment (deivid-rodriguez) [inline .github/workflows/windows-bundler.yml:12]: :scissors:

**Comment by hsbt** [comment]
We should test bundler with rubygems changes always. After that, We will miss regression RubyGems for bundler. 

I totally against to merge this.

**Comment by MSP-Greg** [comment]
@hsbt

> We should test bundler with rubygems changes always.

Maybe you should review the PR again.  It does exactly that.

**Comment by hsbt** [comment]
Thanks, I got it. But We should keep the the reverse pattern too. If We break the RubyGems with bundler changes, We will miss it.

@deivid-rodriguez Do you have a strong reason to merge this?

File: .github/workflows/macos.yml [hunk 7]
```diff
@@ -2,10 +2,17 @@ name: macos
 
 on:
   pull_request:
+    paths-ignore:
+      - "bundler/**"
+      - ".github/workflows/*bundler*.yml"
```
Comment (deivid-rodriguez) [inline .github/workflows/macos.yml:7]: Please also remove all these changes, I didn't suggest to skip any `rubygems` builds. Rubygems tests load the local `bundler` sometimes, and rubygems tests are fast anyways.

**Comment by deivid-rodriguez** [comment]
@hsbt If everything is removed, except from what I suggested initially, I think it's fine. But this PR has a lot of changes I didn't suggest.

**Comment by hsbt** [comment]
Okay, I will review again after that.

**Comment by MSP-Greg** [comment]
@hsbt

> If We break the RubyGems with bundler changes, We will miss it.

I think that's accounted for.  There are a few workflows without any filtering, those are the ones that run with changes to either Bundler or RubyGems.  If I missed something, please let me know.

> I totally against to merge this.

I'd suggest not phrasing things that way.  A question is much better.  If I'm volunteering my time...

**Comment by deivid-rodriguez** [comment]
Closing in favor of #3776.
